utilities/utils/response_unix.go

```diff
@@ -4,8 +4,65 @@ package utils
 
 import (
 	"github.com/docker/docker/api/types"
+	"github.com/patrickmn/go-cache"
+	"github.com/rancher/agent/utilities/config"
+	"github.com/rancher/agent/utilities/constants"
+	"github.com/vishvananda/netlink"
+	"github.com/vishvananda/netns"
+	"time"
 )
 
-func getIP(inspect types.ContainerJSON) string {
-	return inspect.NetworkSettings.IPAddress
+const (
+	cniLabels = "io.rancher.network.cni"
+	linkName  = "eth0"
+)
+
+func getIP(inspect types.ContainerJSON, c *cache.Cache) (string, error) {
+	if val, ok := inspect.Config.Labels[cniLabels]; ok && val == "true" && c != nil {
+		done := make(chan bool, 1)
+		ip := ""
+		go func() {
```

Comment:
This code doesn't need a goroutine from what I see.  Basically you just need the following synchronous logic

```go
ip := get_from_cache()
if ip == "error" {
  return err
}
if ip != "" {
  return ip
}
ip, err = lookup_ip()
if err != nil {
  cache.Put(err)
  return err
}
cache.Put(ip)
return ip
```

Make sure that `lookup_ip` just tries for 1 minutes then the returns an error.  But again you don't need a goroutine for that.  Just a simple loop that tries the lookup, if it fails sleeps a bit, and then if the total time is greater than 1 minute return error otherwise redo the loop.